File: sdk/typescript/src/agent.ts
import type { TradingClient } from './client.js'
import { WatchStream } from './ws.js'
import type { Context, OrderIntent, Strategy } from './strategies.js'
import type { LiveFrame } from './types.js'

export interface RunOptions {
  durationMs?: number
  symbol?: string
  /** Poll REST instead of WebSocket (default false). */
  useWs?: boolean
  log?: (msg: string) => void
}

export interface RunStats {
  ticks: number
  orders: number
  rejected: number
}

/** Binds an agent id to a client and runs strategies against live frames. */
export class Agent {
  ordersSent = 0

  private constructor(
    readonly client: TradingClient,
    readonly agentId: string,
    readonly name: string,
  ) {}

  static async create(client: TradingClient, name: string): Promise<Agent> {
    const created = await client.createAgent(name)
    return new Agent(client, created.agent_id, name)
  }

  submit(intent: OrderIntent) {
    return this.client.placeOrder({
      agent_id: this.agentId,
      symbol: intent.symbol,
      side: intent.side,
      kind: intent.kind,
      qty: intent.qty,
      price: intent.price,
    })
  }

  async run(strategy: Strategy, opts: RunOptions = {}): Promise<RunStats> {
    const {
      durationMs = 60_000,
      symbol = 'NOVA',
      useWs = true,
      log = (m: string) => console.log(m),
    } = opts
    const deadline = Date.now() + durationMs
    const stats: RunStats = { ticks: 0, orders: 0, rejected: 0 }

    const handleFrame = async (frame: LiveFrame): Promise<void> => {
      if (Date.now() > deadline) return
      const desk = await this.client.agent(this.agentId)
      const mandates = frame.mandates ?? []
      const mandate =
        mandates.find((m) => m.agent_id === this.agentId) ?? desk.mandate ?? undefined

      let queued: OrderIntent[] = []
      const ctx: Context = {
        client: this.client,
        agentId: this.agentId,
        frame,
        desk,
        welfare: frame.welfare,
        mandate,
        tournament: frame.tournament ?? null,
        submit: (intent) => queued.push(intent),
      }
      const returned = (strategy.onTick(ctx) as OrderIntent[] | void) ?? []
      for (const intent of [...queued, ...returned]) {
        try {
          await this.submit(intent)
          stats.orders += 1
          this.ordersSent += 1
        } catch (err) {
          stats.rejected += 1
          log(`  order rejected: ${err instanceof Error ? err.message : String(err)}`)
        }
      }
      stats.ticks += 1
      const t = frame.tournament
      log(
        `[${strategy.name}] tick=${stats.ticks} equity=$${desk.equity.toLocaleString('en-US', { maximumFractionDigits: 0 })} role=${mandate?.role ?? '?'} orders=${stats.orders}` +
          (t ? ` | ${t.name}:${t.status}(${t.ticks_left}t)` : ''),
      )
    }

    let stream: WatchStream | null = null
    try {
      if (useWs) {
        stream = new WatchStream(this.client.base, {
          symbol,
          agentId: this.agentId,
          onFrame: (f) => void handleFrame(f),
        }).start()
        while (Date.now() < deadline) await sleep(250)
      } else {
        while (Date.now() < deadline) {
          const snap = await this.client.snapshot(symbol)
          const w = await this.client.welfare().catch(() => ({ agents: [] as never[] }))
          await handleFrame({ ...snap, type: 'snapshot', mandates: w.agents })
          await sleep(1200)
        }
      }
    } finally {
      stream?.close()
    }
    return stats
  }
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms))

File: sdk/trading_engine/agent.py
"""Strategies and the runner that turns live frames into orders."""

from __future__ import annotations

import abc
import time
from dataclasses import dataclass, field
from typing import Any

from .client import TradingClient
from .ws import WatchStream


@dataclass
class OrderIntent:
    symbol: str
    side: str  # "buy" | "sell"
    qty: int
    kind: str = "limit"  # "limit" | "market"
    price: float | None = None


@dataclass
class Context:
    """Everything a strategy may look at for one tick."""

    client: TradingClient
    agent_id: str
    frame: dict[str, Any]
    desk: dict[str, Any]
    welfare: dict[str, Any]
    mandate: dict[str, Any] | None
    tournament: dict[str, Any] | None
    _intents: list[OrderIntent] = field(default_factory=list)

    @property
    def suggestion(self) -> dict[str, Any] | None:
        return (self.mandate or {}).get("suggestion")

    def submit(self, intent: OrderIntent) -> None:
        self._intents.append(intent)

    def _drain(self) -> list[OrderIntent]:
        out = self._intents
        self._intents = []
        return out


class Strategy(abc.ABC):
    """Implement on_tick(); return intents or ctx.submit() them."""

    name: str = "custom"

    @abc.abstractmethod
    def on_tick(self, ctx: Context) -> list[OrderIntent] | None:
        ...


class MandateStrategy(Strategy):
    """The cooperative reference bot: obey the welfare mandate verbatim."""

    name = "mandate"

    def on_tick(self, ctx: Context) -> list[OrderIntent] | None:
        s = ctx.suggestion
        if not s:
            return []
        return [OrderIntent(s["symbol"], s["side"], int(s["qty"]), "limit", float(s["limit"]))]


class GreedyMomentumStrategy(Strategy):
    """The foil: chase momentum, never read the mandate.

    Market-buys symbols that dipped more than 0.5% below prev_close and
    market-sells holdings that ran more than 0.5% above it.
    """

    name = "greedy"

    def __init__(self, clip_qty: int = 10) -> None:
        self.clip_qty = clip_qty

    def on_tick(self, ctx: Context) -> list[OrderIntent] | None:
        intents: list[OrderIntent] = []
        free_cash = float(ctx.desk.get("free_cash", 0.0))
        positions = {p["symbol"]: p for p in ctx.desk.get("positions", [])}
        for stock in ctx.frame.get("stocks", []):
            last = stock.get("last_trade")
            if last is None:
                continue
            change = last / stock["prev_close"] - 1.0
            sym = stock["symbol"]
            pos = positions.get(sym)
            held = int(pos.get("free", 0)) if pos else 0
            if change < -0.005 and free_cash > last * self.clip_qty:
                intents.append(OrderIntent(sym, "buy", self.clip_qty, "market"))
                free_cash -= last * self.clip_qty
            elif change > 0.005 and held > 0:
                intents.append(OrderIntent(sym, "sell", min(held, self.clip_qty), "market"))
        return intents


STRATEGIES = {"mandate": MandateStrategy, "greedy": GreedyMomentumStrategy}


class Agent:
    """Binds an agent id to a client and runs strategies against live frames."""

    def __init__(self, client: TradingClient, agent_id: str, name: str) -> None:
        self.client = client
        self.agent_id = agent_id
        self.name = name
        self.orders_sent = 0

    @classmethod
    def create(cls, client: TradingClient, name: str) -> Agent:
        created = client.create_agent(name)
        return cls(client, created["agent_id"], name)

    def submit(self, intent: OrderIntent) -> dict:
        result = self.client.place_order(
            self.agent_id,
            intent.symbol,
            intent.side,
            intent.kind,
            intent.qty,
            intent.price,
        )
        self.orders_sent += 1
        return result

    def run(
        self,
        strategy: Strategy,
        duration_s: float = 60.0,
        symbol: str = "NOVA",
        use_ws: bool = True,
        log: Any = None,
    ) -> dict[str, Any]:
        """Drive `strategy` from live frames (WS) or polling until duration elapses."""
        printer = log or (lambda msg: print(msg, flush=True))
        deadline = time.monotonic() + duration_s
        stats: dict[str, Any] = {"ticks": 0, "orders": 0}

        def handle_frame(frame: dict[str, Any]) -> None:
            if time.monotonic() > deadline:
                return
            desk = self.client.agent(self.agent_id)
            mandates = frame.get("mandates") or []
            mandate = next((m for m in mandates if m.get("agent_id") == self.agent_id), None)
            ctx = Context(
                client=self.client,
                agent_id=self.agent_id,
                frame=frame,
                desk=desk,
                welfare=frame.get("welfare", {}),
                mandate=mandate,
                tournament=frame.get("tournament"),
            )
            returned = strategy.on_tick(ctx) or []
            intents = ctx._drain() + returned
            rejected = 0
            for intent in intents:
                try:
                    self.submit(intent)
                    stats["orders"] += 1
                except Exception as err:  # noqa: BLE001
                    rejected += 1
                    printer(f"  order rejected: {err}")
            stats["ticks"] += 1
            role = (mandate or {}).get("role", "?")
            t = frame.get("tournament")
            tourney = f" | {t['name']}:{t['status']}({t.get('ticks_left')}t)" if t else ""
            printer(
                f"[{strategy.name}] tick={stats['ticks']} equity=${desk.get('equity', 0):,.0f} "
                f"role={role} orders={stats['orders']}"
                + (f" rejected={rejected}" if rejected else "")
                + tourney
            )

        stream: WatchStream | None = None
        try:
            if use_ws:
                stream = WatchStream(
                    self.client.base_url,
                    symbol=symbol,
                    agent_id=self.agent_id,
                    on_frame=handle_frame,
                ).start()
            while time.monotonic() < deadline:
                if stream is None:
                    snap = self.client.snapshot(symbol)
                    handle_frame({**snap, "mandates": self.client.welfare()["agents"]})
                    time.sleep(1.2)
                else:
                    time.sleep(0.25)
        finally:
            if stream is not None:
                stream.close()
        return stats
